File: src/main/java/org/gecko/view/toolbar/ToolBarBuilder.java
package org.gecko.view.toolbar;

import javafx.geometry.Orientation;
import javafx.scene.control.*;
import javafx.scene.layout.Priority;
import javafx.scene.layout.VBox;
import org.gecko.actions.Action;
import org.gecko.actions.ActionManager;
import org.gecko.tools.Tool;
import org.gecko.tools.ToolType;
import org.gecko.view.ResourceHandler;
import org.gecko.view.views.EditorView;
import org.gecko.view.views.shortcuts.Shortcuts;
import org.gecko.viewmodel.EditorViewModel;
import org.kordamp.ikonli.javafx.FontIcon;

import java.util.List;

import static org.kordamp.ikonli.materialdesign2.MaterialDesignR.REDO;
import static org.kordamp.ikonli.materialdesign2.MaterialDesignU.UNDO;

/**
 * Represents a builder for the {@link ToolBar} displayed in the view, containing a {@link ToggleGroup} with
 * {@link ToggleButton}s for each of the current view's available {@link Tool}s, as well as {@link ToggleButton}s for
 * running the undo and redo operations. Holds a reference to the built {@link ToolBar} and the current
 * {@link EditorView}.
 */
public class ToolBarBuilder {

    private static final String DEFAULT_TOOLBAR_ICON_STYLE_NAME = "toolbar-icon";
    private static final String UNDO_ICON_STYLE_NAME = "undo-toolbar-icon";
    private static final String REDO_ICON_STYLE_NAME = "redo-toolbar-icon";
    private static final int BUTTON_SIZE = 30;

    private final ToolBar toolBar = new ToolBar();
    private final EditorView editorView;

    public ToolBarBuilder(ActionManager actionManager, EditorView editorView, EditorViewModel editorViewModel) {
        this.editorView = editorView;
        toolBar.setOrientation(Orientation.VERTICAL);

        ToggleGroup toggleGroup = new ToggleGroup();

        toggleGroup.selectedToggleProperty().addListener((observable, oldToggle, newToggle) -> {
            if (newToggle == null) {
                toggleGroup.selectToggle(oldToggle);
            }
        });

        for (int i = 0; i < editorViewModel.getTools().size(); i++) {
            addTools(actionManager, toggleGroup, editorViewModel.getTools().get(i));

            // add separator
            if (i < editorViewModel.getTools().size() - 1) {
                toolBar.getItems().add(new Separator());
            }
        }

        // Undo and Redo buttons
        toolBar.getItems().add(new Separator());
        VBox spacer = new VBox();
        VBox.setVgrow(spacer, Priority.ALWAYS);

        toolBar.getItems().add(spacer);

        Button undoButton = new Button(ResourceHandler.getString("Buttons", "undo"),
                FontIcon.of(UNDO, 24));

        String toolTip = "%s (%s)".formatted(
                ResourceHandler.getString("Tooltips", "undo"),
                Shortcuts.UNDO.get().getDisplayText());

        undoButton.setTooltip(new Tooltip(toolTip));
        undoButton.setOnAction(event -> actionManager.undo());
        //undoButton.getStyleClass().add(DEFAULT_TOOLBAR_ICON_STYLE_NAME);
        //undoButton.getStyleClass().add(UNDO_ICON_STYLE_NAME);

        var redoButton = new Button(ResourceHandler.getString("Buttons", "redo"),
                FontIcon.of(REDO, 24));
        toolTip =
                "%s (%s)".formatted(ResourceHandler.getString("Tooltips", "redo"), Shortcuts.REDO.get().getDisplayText());
        redoButton.setTooltip(new Tooltip(toolTip));
        redoButton.setOnAction(event -> actionManager.redo());
        redoButton.setContentDisplay(ContentDisplay.GRAPHIC_ONLY);
        undoButton.setContentDisplay(ContentDisplay.GRAPHIC_ONLY);
        toolBar.getItems().addAll(undoButton, redoButton);
    }

    private void addTools(ActionManager actionManager, ToggleGroup toggleGroup, List<Tool> toolList) {
        for (Tool tool : toolList) {
            ToolType toolType = tool.getToolType();
            ToggleButton toolButton = new ToggleButton(
                    toolType.getLabel(),
                    FontIcon.of(toolType.getIcon(),24));
            toolButton.getStyleClass().add(DEFAULT_TOOLBAR_ICON_STYLE_NAME);

            //Would like to bind the selectedproperty of the button here but cannot because of a javafx bug
            editorView.getViewModel().getCurrentToolProperty().addListener((observable, oldValue, newValue) ->
                    toolButton.setSelected(newValue == tool));
            toolButton.setSelected(editorView.getViewModel().getCurrentToolType() == toolType);

            toolButton.selectedProperty().addListener((observable, oldValue, newValue) -> {
                if (newValue) {
                    Action action = actionManager.getActionFactory().createSelectToolAction(toolType);
                    actionManager.run(action);
                }
            });

            //toolButton.getStyleClass().add(toolType.getIcon());
            Tooltip tooltip = new Tooltip("%s (%s)"
                    .formatted(toolType.getLabel(), toolType.getKeyCodeCombination().getDisplayText()));
            toolButton.setTooltip(tooltip);
            toolBar.getItems().add(toolButton);
            toggleGroup.getToggles().add(toolButton);
        }
    }

    public ToolBar build() {
        return toolBar;
    }
}

File: src/main/kotlin/org/gecko/view/toolbar/ToolBarBuilder.kt
package org.gecko.view.toolbar

import javafx.beans.value.ObservableValue
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Orientation
import javafx.scene.control.*
import javafx.scene.layout.FlowPane
import javafx.scene.layout.Priority
import javafx.scene.layout.VBox
import org.gecko.actions.Action
import org.gecko.actions.ActionManager
import org.gecko.tools.Tool
import org.gecko.view.ResourceHandler
import org.gecko.view.views.EditorView
import org.gecko.view.views.shortcuts.Shortcuts
import org.gecko.viewmodel.EditorViewModel
import org.kordamp.ikonli.javafx.FontIcon
import org.kordamp.ikonli.materialdesign2.MaterialDesignR
import org.kordamp.ikonli.materialdesign2.MaterialDesignU

/**
 * Represents a builder for the [ToolBar] displayed in the view, containing a [ToggleGroup] with
 * [ToggleButton]s for each of the current view's available [Tool]s, as well as [ToggleButton]s for
 * running the undo and redo operations. Holds a reference to the built [ToolBar] and the current
 * [EditorView].
 */
class ToolBarBuilder(
    actionManager: ActionManager,
    val editorView: EditorView,
    editorViewModel: EditorViewModel
) {
    val toolBar = FlowPane() //ToolBar()

    init {
        toolBar.orientation = Orientation.VERTICAL

        val toggleGroup = ToggleGroup()

        toggleGroup.selectedToggleProperty()
            .addListener { observable: ObservableValue<out Toggle?>?, oldToggle: Toggle?, newToggle: Toggle? ->
                if (newToggle == null) {
                    toggleGroup.selectToggle(oldToggle)
                }
            }

        for (i in editorViewModel.tools.indices) {
            addTools(actionManager, toggleGroup, editorViewModel.tools[i])

            // add separator
            if (i < editorViewModel.tools.size - 1) {
                toolBar.children.add(Separator())
            }
        }

        // Undo and Redo buttons
        toolBar.children.add(Separator())
        val spacer = VBox()
        VBox.setVgrow(spacer, Priority.ALWAYS)

        toolBar.children.add(spacer)

        val undoButton = Button(
            ResourceHandler.Companion.undo,
            FontIcon.of(MaterialDesignU.UNDO, 24)
        )

        var toolTip = "%s (%s)".format(
            ResourceHandler.Companion.undo,
            Shortcuts.UNDO.get().displayText
        )

        undoButton.tooltip = Tooltip(toolTip)
        undoButton.onAction = EventHandler { event: ActionEvent? -> actionManager.undo() }

        //undoButton.getStyleClass().add(DEFAULT_TOOLBAR_ICON_STYLE_NAME);
        //undoButton.getStyleClass().add(UNDO_ICON_STYLE_NAME);
        val redoButton = Button(ResourceHandler.redo, FontIcon.of(MaterialDesignR.REDO, 24))
        toolTip = String.format("%s (%s)", ResourceHandler.redo, Shortcuts.REDO.get().displayText)
        redoButton.tooltip = Tooltip(toolTip)
        redoButton.onAction = EventHandler { event: ActionEvent? -> actionManager.redo() }
        redoButton.contentDisplay = ContentDisplay.GRAPHIC_ONLY
        undoButton.contentDisplay = ContentDisplay.GRAPHIC_ONLY
        //toolBar.children.addAll(undoButton, redoButton)
    }

    fun addTools(actionManager: ActionManager, toggleGroup: ToggleGroup, toolList: List<Tool>) {
        for (tool in toolList) {
            val toolType = tool.toolType
            val toolButton = ToggleButton(
                toolType.label,
                FontIcon.of(toolType.icon, 24)
            )
            toolButton.styleClass.add(DEFAULT_TOOLBAR_ICON_STYLE_NAME)

            //Would like to bind the selectedproperty of the button here but cannot because of a javafx bug
            editorView.viewModel.currentToolProperty.addListener { observable: ObservableValue<out Tool>?, oldValue: Tool?, newValue: Tool ->
                toolButton.isSelected = newValue === tool
            }
            toolButton.isSelected = editorView.viewModel.currentToolType == toolType

            toolButton.selectedProperty()
                .addListener { observable: ObservableValue<out Boolean>?, oldValue: Boolean?, newValue: Boolean ->
                    if (newValue) {
                        val action: Action = actionManager.actionFactory.createSelectToolAction(toolType)
                        actionManager.run(action)
                    }
                }

            //toolButton.getStyleClass().add(toolType.getIcon());
            val tooltip = Tooltip("%s (%s)".format(toolType.label, toolType.keyCodeCombination?.displayText))
            toolButton.tooltip = tooltip
            toolBar.children.add(toolButton)
            toggleGroup.toggles.add(toolButton)
        }
    }

    fun build(): javafx.scene.Node {
        return toolBar
    }

    companion object {
        const val DEFAULT_TOOLBAR_ICON_STYLE_NAME = "toolbar-icon"
        const val UNDO_ICON_STYLE_NAME = "undo-toolbar-icon"
        const val REDO_ICON_STYLE_NAME = "redo-toolbar-icon"
        const val BUTTON_SIZE = 30
    }
}
